core/src/main/scala/org/apache/spark/scheduler/ShuffleMapTask.scala

```diff
@@ -61,28 +62,52 @@ private[spark] class ShuffleMapTask(
     // Deserialize the RDD using the broadcast variable.
     val deserializeStartTime = System.currentTimeMillis()
     val ser = SparkEnv.get.closureSerializer.newInstance()
-    val (rdd, dep) = ser.deserialize[(RDD[_], ShuffleDependency[_, _, _])](
+    val (rdd, dep) = ser.deserialize[(RDD[_], BaseShuffleDependency[_])](
       ByteBuffer.wrap(taskBinary.value), Thread.currentThread.getContextClassLoader)
     _executorDeserializeTime = System.currentTimeMillis() - deserializeStartTime
-
     metrics = Some(context.taskMetrics)
-    var writer: ShuffleWriter[Any, Any] = null
-    try {
-      val manager = SparkEnv.get.shuffleManager
-      writer = manager.getWriter[Any, Any](dep.shuffleHandle, partitionId, context)
-      writer.write(rdd.iterator(partition, context).asInstanceOf[Iterator[_ <: Product2[Any, Any]]])
-      writer.stop(success = true).get
-    } catch {
-      case e: Exception =>
+
+    dep match {
+      case objDep: ShuffleDependency[_, _, _] =>
+        var writer: ShuffleWriter[Any, Any] = null
+        try {
+          val manager = SparkEnv.get.shuffleManager
+          writer = manager.getWriter[Any, Any](objDep.shuffleHandle, partitionId, context)
+          writer.write(rdd.iterator(partition, context)
+            .asInstanceOf[Iterator[_ <: Product2[Any, Any]]])
+          writer.stop(success = true).get
+        } catch {
+          case NonFatal(e) =>
+            try {
+              if (writer != null) {
+                writer.stop(success = false)
+              }
+            } catch {
+              case NonFatal(e2) =>
+                log.debug("Could not stop writer", e2)
+            }
+            throw e
+        }
+      case binaryDep: BinaryShuffleDependency[_] =>
+        // TODO(josh): Work around this limitation / document it.
```

Comment:
If we want to expose a super-low-level API for shuffle write, files, then the file format becomes part of the API contract. Specifically, this interface is restricting writers to write data in a sort-shuffle-compatible format. If the hash-based shuffle manager is enabled, then this custom write path will not work.

There are multiple ways that we could address this limitation. One approach might be to remove hash-based shuffle. Another might be to refactor the shuffle block resolver mechanism such that a single Spark application can serve shuffle files generated in either format.